readme.md
# Prerequisites:
- Raspberry Pi 4 (RPI4). It is tested on RPI4, but should also support Raspberry Pi Compute Module 4 or Raspberry Pi 3.
- Raspberry Pi OS Bullseye 64bit (recommended) or 32bit. See compilation notes below to set the OS version.
- Mira220 sensor board V3.0 or Mira050 sensor board V1.0.
## If cross-compiling with a host x86 machine
- Host machine running Ubuntu 18.04 or later.
- Host has the tool chain needed to compile a standard Raspberry Pi Linux Kernel. Issue this command `sudo apt install git bc bison flex libssl-dev make libc6-dev libncurses5-dev crossbuild-essential-armhf crossbuild-essential-arm64`. Refer to [Raspberry Pi doc](https://www.raspberrypi.com/documentation/computers/linux_kernel.html#cross-compiling-the-kernel) for further details. Preferably, make sure the host has the required tool by compiling a standard Raspberry Pi kernel as stated in the Raspberry Pi doc.
## If compiling natively on the raspberry pi
- The Raspberry Pi has 64-bit Desktop version of Raspberry Pi OS image [link](https://downloads.raspberrypi.org/raspios_arm64/images/raspios_arm64-2022-09-07/), or (not recommended) 32-bit version [link](https://downloads.raspberrypi.org/raspios_full_armhf/images/raspios_full_armhf-2022-09-07/).
- The Raspberry Pi has the required tools. Log into the RPI and issue this command `sudo apt install git bc bison flex libssl-dev make`. Refer to [Raspberry Pi doc](https://www.raspberrypi.com/documentation/computers/linux_kernel.html#building-the-kernel-locally) for further details.

# Compilation and installation:
## If cross-compiling with a host x86 machine
- Clone or copy this repo to an x86 machine. Make sure the host machine has prerequisites mentioned in the previous section.
- Before compilation, configure whether 64bit OS or 32bit OS is built. Open the file `common/build.sh` and edit the variable `OSBIT` to be `64` or `32`. This variable controls a few build options. For example, it controls the `KERNEL` option that points to the actual kernel file used in the RPI. By default, 32-bit RPI 4 uses `kernel7l.img`, 64-bit RPI 4 uses `kernel8.img`.
- Before compilation, configure the RPI hardware type. Open the file `common/build.sh` and edit the variable `RPIHW` to be `RPI4B` or `RPI3B`. This variable controls what kind of build config is used.
- Open the file `common/build.sh` and point `RPI_KERNEL_INSTALL_DIR=` to the actual location that the Raspberry Pi SD card is mounted on the Ubuntu host. Alternative to mounting the SD card by physically plugging it on to the Ubuntu host, it is possible to mount it via ssh by the command `sudo sshfs -o sftp_server="/usr/bin/sudo /usr/lib/openssh/sftp-server" pi@IP_OF_PI:/ /media/pi` where `IP_OF_PI` is the IP address of pi and `/media/pi` is an example mount point on the host.
- Run `build_all.sh`. The script performs the following: (1) Download a specific version of Linux kernel source; (2) Apply Mira220 and Mira050 related source code and patches; (3) Prepare compilation script; (4) Performs actual compilation and installation. To re-run the `build_all.sh` for a specific step, comment out the parts that are not relevant.
## If compiling natively on the raspberry pi
- Clone or copy this git repo to the target RPI. Make sure the RPI has the prerequisites mentioned in the previous section.
- Run `build_native.sh`. It builds the drivers and device trees into the Debian package, and then installs the Debian package to the local RPI.
- The Debian package, with a name like `mira-driver_0.1.5-1_arm64.deb`, is built for a specific architecture (check via `dpkg --print-architecture`) and a specific kernel version (check via `uname -r`) of the RPI. If other RPIs have the same architecture and the same kernel version, users can simply copy the Debian package over to other RPIs and install the driver by two commands: `sudo dpkg -i mira-driver_0.1.5-1_arm64.deb` for installation and then `sudo depmod` for post-installation update.

## Configuration
- Post-installation, log on to the Raspberry Pi, add a new line to `/boot/config.txt`. Depending on whether Mira220 mono or Mira220 color or Mira050 mono is connected, this new line will be either `dtoverlay=mira220` for Mira220 mono, or `dtoverlay=mira220color` for Mira220 color, or `dtoverlay=mira050` for Mira050 mono (pick one and only one!). This line tells the RPI to load the corresponding driver at boot time.
- Reboot to let the configuration take effect.

# Tests:
- Test whether installation script successfully copies files. After reboot, issue the command `uname -a` and the printed out timestamp of the kernel should match the time that you compile it.
- Test whether Mira220 or Mira050 driver is loaded. As mentioned above, the line `dtoverlay=...` should be added to `/boot/config.txt` on RPI. After reboot, the `dmesg` command should print out logs with keywords `MIRA220` and `MIRA220PMIC` for Mira220 driver; `MIRA050` and `MIRA050PMIC` for Mira050 driver;. For a 64 bit kernel - it could be that you need to add `kernel=kernel8.img` to `config.txt`
- Test whether the power management IC driver module (MIRA220PMIC/MIRA050PMIC) is working. The green LED on the sensor board should be turned on.
- To further test the actual driver module (MIRA220/MIRA050), please refer to a separate repo `ams_rpi_software` and follow instructions from there.

# Post-installation:
- Install other custom driver modules or software if needed. For example, the Quadric Dev Kit driver (`thor`) is located in a separate repo [link](https://gittf.ams-osram.info/cis_solutions/raspberry_evk/quadric_driver).
- Instructions on creating a custom OS image from a plain OS image are described in [doc/create_os_image.md](doc/create_os_image.md).

# Release workflow
Two repositories, `ams_rpi_kernel` and `ams_rpi_software`, are released. Assumption is that, both repositories have a git tag for the release version, such as `v0.1.2`. Below are commands that create a specific release from the tag.
```
# Export the release tag version, for example, v0.1.2
export RELEASE_TAG=v0.1.2
# Create release pacakge ams_rpi_kernel_${RELEASE_TAG}.tar.gz
cd ams_rpi_kernel
git archive --prefix=ams_rpi_kernel/ -o ams_rpi_kernel_${RELEASE_TAG}.tar.gz ${RELEASE_TAG}
# Create release for ams_rpi_software_${RELEASE_TAG}.tar.gz
cd ams_rpi_software
git archive --prefix=ams_rpi_software/ -o ams_rpi_software_${RELEASE_TAG}.tar.gz ${RELEASE_TAG}
```


doc/release_notes.md
# Version v0.1.11

Bug fixes:
- Fix driver code to compile with latest RPI Linux kernel and tags used by pi-gen.

Known issues:
- From now on, Mira driver is no longer compatible to Linux kernel < 6.1.21.

# Version v0.1.10

New features:
- Add LED control to Mira050 driver.

Bug fixes:
- Fix Mira050 driver duplicated modes and wrong fps.

# Version v0.1.9

Bug fixes:
- Fix bug when integrating latest libcamera tag v0.0.2.

# Version v0.1.7 and v0.1.8

New features:
- Enable 8/10/12 bit modes for Mira050.
- Mira050 12 bit mode uses coarse gains 1x/2x/4x.
- Mira050 8 and 10 bit modes use fine gain.

Bug fixes:
- Fix --list-cameras bug on Mira050.

# Version v0.1.6

New features:
- Bump to `raspios_arm64-2022-09-07`, kernel 5.15.61-v8+, tag 1.20220830.
- Add script `build_native.sh` to build the driver natively on an RPI.

Bug fixes:
- Fix module probe order problem by including pmic module inside mira220/050 module.
- Fix over exposure problem of Mira220, and strange exposure of Mira050.


# Version v0.1.5

New features:
- Support mira220color and mira050color.
- Mira220 driver supports two modes: (1) 640x480 120fps; (2) 1600x1400 30fps. BPP:8/10/12.

Bug fixes:
- Fix Mira220 bring up problem by increasing reset time.

# Version v0.1.4

New features:
- Add Mira050 support. Reduce Mira220 power consumption by keeping reset signal high most of the time.

# Version v0.1.3

Bug fixes:
- Increase reset timing to avoid instable reset.

# Version v0.1.2

New features:
- Use simplified flags `OSBIT` and `RPIHW` in `common/build.sh` to configure 32&64bit OS, and HW version of RPI.

Bug fixes:
- The RPI Linux kernel tag rpi-5.15.y is unstable. Use commit id instead. Tested on 32&64bit OS.

# Version v0.1.1

New features:
- Add 64-bit OS support (tested by overwriting an 32bit OS).

# Version v0.1.0

New features:
- Basic mode: 1600x1400 with 8 bits per pixel.

Known issues:
- Each time a new image or video is captured, the mira220pmic module requires reloading.



common/build.sh
#!/usr/bin/env bash

set -e

cd "$( dirname "${BASH_SOURCE[0]}" )"

# Set OSBIT to 32 or 64 to switch builds
OSBIT=64

# SET RPIHW to RPI4B or RPI3BP
RPIHW=RPI4B

# The OSBIT and RPIHW controls KERNEL and CONFIG
# Details in doc:
# https://www.raspberrypi.com/documentation/computers/linux_kernel.html#kernel-configuration

# Configure variables according to OSBIT and RPIHW
if [ "$OSBIT" = "32" ]; then
	# For 32-bit OS, use the following settings
	export ARCH=arm
	export CROSS_COMPILE=arm-linux-gnueabihf-
	export IMG_NAME=zImage
	if [ "$RPIHW" = "RPI4B" ]; then
		export KERNEL=kernel7l
	elif [ "$RPIHW" = "RPI3B" ]; then
		export KERNEL=kernel7
	else
		echo "ERROR: RPIHW should be RPI4B or RPI3B."
		exit 1
	fi
elif [ "$OSBIT" = "64" ]; then
	# For 64-bit OS, use the following settings instead
	export ARCH=arm64
	export CROSS_COMPILE=aarch64-linux-gnu-
	export IMG_NAME=Image
	export KERNEL=kernel8
else
	echo "ERROR: OSBIT should be 32 or 64."
	exit 1
fi

if [ "$RPIHW" = "RPI4B" ]; then
	CONFIG=bcm2711_defconfig
elif [ "$RPIHW" = "RPI3B" ]; then
	if [ "$OSBIT" = "64" ]; then
		CONFIG=bcm2711_defconfig
	elif [ "$OSBIT" = "32" ]; then
		CONFIG=bcm2709_defconfig
	else
		echo "ERROR: OSBIT should be 32 or 64."
		exit 1
	fi
else
	echo "ERROR: RPIHW should be RPI4B or RPI3B."
	exit 1
fi

#
# The variables below are independent of OSBIT and RPIHW
#

NAME=$0
COMMAND=$1
INSTALL_DIR=${RPI_KERNEL_INSTALL_DIR:-/media/pi}
BUILD_DIR=${RPI_KERNEL_BUILD_DIR:-build}

# ASCII Escape sequences
RED_COLOR=31
YELLOW_COLOR=33
BLUE_COLOR=34
CYAN_COLOR=36

set_attribute() {
    echo -ne "\u001b[${1}m"
}

reset_attributes() {
    echo -ne "\u001b[0m"
}

log_warning() {
    set_attribute ${YELLOW_COLOR}
    echo "$@"
    reset_attributes
}

log_error() {
    set_attribute ${RED_COLOR}
    echo "$@"
    reset_attributes
}

usage() {
    set_attribute ${BLUE_COLOR}
    cat <<EOF
Usage: ${NAME} COMMAND
Supported commands are:
    clean:            Cleans the kernel output.
    restoreconfig:    Creates the config from the ${CONFIG} target.
    menuconfig:       Configure the target interactively using ncurses.
    build:            Builds kernel image, modules and device tree binaries.
    modules:          Builds the kernel modules only.
    dtbs:             Builds the device tree binaries only.
    kernel:           Builds the compressed kernel image only.
    install:          Builds and installs kernel and modules in the rootdir specified by
                      RPI_KERNEL_INSTALL_DIR.
    install_modules:  Builds and installs modules in the rootdir specified by RPI_KERNEL_INSTALL_DIR.
    install_dtbs:     Builds and installs dtbs in the rootdir specified by RPI_KERNEL_INSTALL_DIR.
    install_kernel:   Builds and installs kernel in the rootdir specified by RPI_KERNEL_INSTALL_DIR.
    compdb:           Generates a compilation database that can be used with clangd or IntelliSense
                      in VSCode.
EOF
    reset_attributes
    exit 1
}

install_cross_compiler() {
    echo "Would you like to install it now? (you may be asked for your password)"
    select response in "Yes" "No"; do
        case ${response} in
            Yes) sudo apt install -y gcc-arm-linux-gnueabihf gcc-aarch64-linux-gnu; break;;
            No) exit 0;;
            *) echo "Please select one of the available options";;
        esac
    done
}

check_dependencies() {
    if ! ${CROSS_COMPILE}gcc --version > /dev/null 2>&1; then
        log_warning "Could not find ${CROSS_COMPILE}gcc, which is required to build the Raspberry Pi kernel."
        install_cross_compiler
    fi
}

run_command() {
    set_attribute ${CYAN_COLOR}
    echo -n "Running command: "
    reset_attributes
    echo "\"$@\""
    if ! "$@"; then
        log_error "Error running command"
        exit 1
    fi
}
run_command_as_su_no_fail() {
    set_attribute ${CYAN_COLOR}
    echo -n "Running command as super user: "
    reset_attributes
    echo "\"$@\""
    if ! "sudo" "$@"; then
        log_error "Error running command"
    fi
}

run_command_as_su() {
    set_attribute ${CYAN_COLOR}
    echo -n "Running command as super user: "
    reset_attributes
    echo "\"$@\""
    if ! "sudo" "$@"; then
        log_error "Error running command"
        exit 1
    fi
}

run_make() {
    run_command make -j"$(nproc)" "$@" INSTALL_MOD_PATH="${INSTALL_DIR}" "O=${BUILD_DIR}"
}

run_install_modules() {
    run_command_as_su make -j"$(nproc)" modules_install INSTALL_MOD_PATH="${INSTALL_DIR}" "O=${BUILD_DIR}"
}

run_install_dtbs() {
    run_command_as_su mkdir -p "${INSTALL_DIR}/boot/overlays"
    if [ "$ARCH" = "arm" ]; then
        run_command_as_su cp "${BUILD_DIR}"/arch/${ARCH}/boot/dts/*.dtb "${INSTALL_DIR}/boot/"
    else
        run_command_as_su cp "${BUILD_DIR}"/arch/${ARCH}/boot/dts/broadcom/*.dtb "${INSTALL_DIR}/boot/"
    fi
    run_command_as_su cp "${BUILD_DIR}"/arch/${ARCH}/boot/dts/overlays/*.dtb* "${INSTALL_DIR}/boot/overlays"
}

run_install_kernel() {
    run_command_as_su mkdir -p "${INSTALL_DIR}/boot"
    run_command_as_su_no_fail if [ -e "${INSTALL_DIR}/boot/$KERNEL.img" ] ; then cp "${INSTALL_DIR}/boot/$KERNEL.img" "${INSTALL_DIR}/boot/${KERNEL}-backup.img"; fi
    run_command_as_su cp "${BUILD_DIR}/arch/${ARCH}/boot/${IMG_NAME}" "${INSTALL_DIR}/boot/${KERNEL}.img"
}

run_install() {
    run_install_modules
    run_install_kernel
    run_install_dtbs
}

run_build() {
    if [ ! -f ${BUILD_DIR}/.config ];
    then
        log_warning "No configuration found, using default configuration."
        run_make ${CONFIG}
    fi
    run_make "$@"
}

gen_compile_commands() {
    KERNEL_SOURCE_DIR=${PWD}
    run_command pushd ${BUILD_DIR}
    run_command ${KERNEL_SOURCE_DIR}/scripts/clang-tools/gen_compile_commands.py
    run_command popd
    run_command ln -sf ${BUILD_DIR}/compile_commands.json compile_commands.json
}

check_dependencies

case ${COMMAND} in
    clean)
        run_make clean
        run_command rm -rf ${BUILD_DIR};;
    restoreconfig) run_make ${CONFIG};;
    menuconfig) run_make menuconfig;;
    build) run_build ${IMG_NAME} dtbs modules;;
    modules) run_build modules;;
    dtbs) run_build dtbs;;
    kernel) run_make ${IMG_NAME};;
    install) run_install;;
    install_modules) run_install_modules;;
    install_dtbs) run_install_dtbs;;
    install_kernel) run_install_kernel;;
    compdb) gen_compile_commands;;
    --) run_make "${@:2}";;
    *) usage;;
esac
